src/System.Net.Http/src/System/Net/Http/Managed/HttpEnvironmentProxy.cs

```diff
@@ -0,0 +1,199 @@
+            }
+        }
+
+        /// <summary>
+        /// This function will evaluate given string and it will try to convert
+        /// it to Uri object. The string could contain URI fragment, IP address and  port
+        /// tuple or just IP address or name. 
+        /// </summary>
+        private Uri GetUriFromString(String value)
+        {
+            if (!value.StartsWith("http://", StringComparison.OrdinalIgnoreCase) &&
+                !value.StartsWith("https://", StringComparison.OrdinalIgnoreCase))
+            {
+                // Default to HTTP
+                value = "http://" + value;
+            }
+            Uri uri = new Uri(value);
+
+            if (uri.Scheme != Uri.UriSchemeHttp && uri.Scheme != Uri.UriSchemeHttps)
+            {
+                throw new NotSupportedException("Only HTTP and HTTPS protocols are supported for proxy.");
+            }
+            return uri;
+        }
+
+        /// <summary>
+        /// This function returns true if given Host mach bypss list.
+        /// Note, that the list is common for http and https.
+        /// </summary>
+        private bool IsMatchInBypassList(Uri input)
+        {
+            if (_bypass != null)
+            {
+                foreach (string s in _bypass)
+                {
+                    if (s[0] == '.')
+                    {
+                        // This should match either domain it self or any subdomain or host
+                        // .foo.com will match foo.com it self or *.foo.com
+                        if ((s.Length - 1) == input.Host.Length &&
+                            String.Compare(s, 1, input.Host, 0, input.Host.Length, StringComparison.OrdinalIgnoreCase) == 0)
+                        {
+                            return true;
+                        }
+                        else if (input.Host.EndsWith(s, StringComparison.OrdinalIgnoreCase))
+                        {
+                                return true;
+                        }
+
+                    }
+                    else
+                    {
+                        if (String.Compare(s, input.Host, StringComparison.OrdinalIgnoreCase) == 0)
+                        {
+                            return true;
+                        }
+                    }
+
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Gets the proxy URI. (iWebProxy interface)
+        /// </summary>
+        public Uri GetProxy(Uri uri)
+        {
+            return uri.Scheme == "http" ? _http : _https;
+        }
+
+        /// <summary>
+        /// CHecks if URI is subject to proxy or not. 
+        /// </summary>
+        public bool IsBypassed(Uri uri)
+        {
+            bool ret =  (uri.Scheme == "http" ? _http : _https) == null;
+
+            if (ret)
+            {
+                return ret;
+            }
+            return IsMatchInBypassList(uri);
+        }
+
+        public ICredentials Credentials
```

Comment:
I think we're going to have to do something more involved for credential handling here.

The Credentials property should return an object that returns the right credential when its ICredential.GetCredential method is called.  That's how the proxy handling code is expecting this to work (I think anyway).

We could use the CredentialCache object here, or we could provide our own ICredential implementation.  

This should make GetGredentials() below unnecessary.

Reply:
The part I don't know how to do @geoffkizer is determination if http or https should be returned.  One can cache the last queried URL but I'm not sure how that would work for multi-threaded app. The current structure seems to assume that there is only single set of credentials for all protocol. Do I miss something?

Reply:
ICredentials.GetCredential takes a Uri.  So, your IWebProxy.Credentials needs to return an ICredentials object that knows about the two proxy Uris and their associated credentials.  Then, when ICredentials.GetCredential() is called, check if the Uri is the http or https proxy Uri and return the appropriate NetworkCredential -- or null if it doesn't match.

Reply:
I updated the code to either return standard credential object of only one protocol has authentication set or if they are equal. We will use new wrapped object if http and https needs different authentication. Please take a look. I plan to add CredentialCache as part of #26405